packages/transaction-metrics-exporter/src/block-processor.ts

```diff
@@ -0,0 +1,266 @@
+    private blockInterval = 1,
+    private fromBlock: number = 0,
+    private toBlock?: number
+  ) {}
+
+  async init() {
+    if (this.initialized) {
+      throw new Error('BlockProcessor is running')
+    }
+    this.initialized = true
+
+    await this.loadContracts()
+
+    if (this.fromBlock?.toFixed?.()) {
+      await this.initBatch()
+      if (!this.toBlock) {
+        await this.initSubscription()
+      }
+    } else {
+      await this.initSubscription()
+    }
+  }
+
+  async loadContracts() {
+    this.contracts.Exchange = await this.kit.contracts.getExchange()
+    this.contracts.SortedOracles = await this.kit.contracts.getSortedOracles()
+    this.contracts.Reserve = await this.kit.contracts.getReserve()
+    this.contracts.GoldToken = await this.kit.contracts.getGoldToken()
+    this.contracts.EpochRewards = await this.kit.contracts.getEpochRewards()
+    this.contracts.StableToken = await this.kit.contracts.getStableToken()
+  }
+
+  async initSubscription() {
+    const subscription = await this.kit.web3.eth.subscribe('newBlockHeaders')
+
+    // Prevent same block multiples times
+    let lastBlocks: number[] = []
+    subscription.on('data', (header) => {
+      if (!lastBlocks.includes(header.number)) {
+        // tslint:disable-next-line: no-floating-promises
+        this.onNewBlock(header.number)
+      }
+      lastBlocks.push(header.number)
+      lastBlocks = lastBlocks.slice(-10)
+    })
+  }
+
+  async initBatch() {
+    let block = this.fromBlock - 1
+    while (++block <= (await this.getToBlock())) {
+      await this.onNewBlock(block, block - this.fromBlock)
+    }
+  }
+
+  async getToBlock() {
+    if (this.toBlock) {
+      return this.toBlock
+    }
+    return (await this.kit.web3.eth.getBlock('latest')).number
+  }
+
+  async onNewBlock(blockNumber: number, blockIntervalNumber?: number) {
+    if ((blockIntervalNumber ?? blockNumber) % this.blockInterval === 0) {
+      await Promise.all([this.fetchBlockState(blockNumber), this.processBlockHeader(blockNumber)])
+    }
+  }
+
+  async fetchBlockState(blockNumber: number) {
+    const promises = stateGetters.map(
+      async ({ contract, method, args, transformValues, maxBucketSize }) => {
+        if (!(await this.contracts[contract].exists(blockNumber))) {
+          return
+        }
+        args = typeof args === 'function' ? args(blockNumber) : args
+        this.contracts[contract].setDefaultBlock(blockNumber)
+        return (this.contracts as any)[contract][method](...args)
+          .then((returnData: any) => {
+            this.contracts[contract].setDefaultBlock('latest')
+            const data: DataResult = {
+              contract,
+              function: method,
+              args: JSON.stringify(args),
+              blockNumber,
+              values: transformValues(returnData),
+            }
+            this.logEvent(LoggingCategory.State, data)
+            this.logHistogram({ ...data, args }, maxBucketSize)
```

Comment:
This means that every function call that we want to do is always going to be logged as a histogram? Is that reasonable?

Reply:
probably the metric format needs to become part of the state view definition / config

@Pedro-vk can you log a new issue for this?

Reply:
@diminator Done: https://github.com/celo-org/celo-monorepo/issues/3542